### PR TITLE
Fix Build System: Push Events and root file changes

### LINKED ISSUE #140: Differential build PR bugs
Need to fix:

- [ ] `python3 stats.py --build --files test` build all solutions and shouldn't.
- [ ] `push` travis events fails on build as [this](https://travis-ci.org/DestructHub/ProjectEuler/builds/301674042?utm_source=github_status&utm_medium=notification)

Related to #137  #139  #136

### PR DESCRIPTION
**Pull request by ryukinix**
This should fix #140. Only merge this when all the issues related to
the #140 has been fixed.

This pull-request is related to PR #142 (merged).

### CODE REVIEW

**Comment by ryukinix** [comment]
On this new way of PR, we'll only allow code be merging after someone approves it explicilty through the GitHub Review System.

So, as I'm the author of this PR now, you should approve that to I can merge it.

File: test [hunk 1]
```diff
@@ -1,5 +1,11 @@
 #!/bin/bash
 
+if [ -v TRAVIS_PULL_REQUEST_BRANCH ]; then
```
Comment (A-Hilaly) [inline test]: This look more logic

Comment (ryukinix) [reply test]: Yes, but seems that `-v` is not sufficient here because is set somewhere as empty. But if I use `-z` instead this can be fixed

Comment (A-Hilaly) [reply test]: I think we need ! -z because -z means is empty

Comment (A-Hilaly) [reply test]: ! -z just skipped both builds ..

**Comment by A-Hilaly** [comment]
I found this [doc](https://docs.travis-ci.com/user/pull-requests/#Security-Restrictions-when-testing-Pull-Requests) talking about TRAVIS_PULL_REQUEST value that should exist in PR case

**Comment by A-Hilaly** [comment]
I think I understand why it don't find any environments, we are running test using bash ... exports doesn't show up cause they need to in `.bash_profile/rc `files to be loaded, ill fix this

**Comment by A-Hilaly** [comment]
@ryukinix I'm sorry I didnt see your commit before

File: test [hunk 2]
```diff
@@ -52,7 +52,7 @@ function main () {
             build_staged
         fi
     else
-        echo "Usage : test.sh [sync|async] [diff|staged]"
+        echo "Usage : ./test [sync|async] [diff|staged]"
```
Comment (ryukinix) [inline test]: use `$0` instead

**Comment by ryukinix** [comment]
> @ryukinix I'm sorry I didnt see your commit before
Please, always use `git pull --rebase` on shared branchs. On set this as default on your `git config`. 

I described the problem here on this commit: https://github.com/DestructHub/ProjectEuler/pull/143/commits/7b79deb1cf86a4ef953e83d3d943f83932608755

File: .travis.yml [hunk 3]
```diff
@@ -1,7 +1,8 @@
 sudo: required
 
--services:
- -  - docker
+services:
+  - docker
 
 script:
+  - sudo chmod +x "$PWD/test"
```
Comment (ryukinix) [inline .travis.yml]: why not just `./test` instead?

File: docker-compose.yml [hunk 4]
```diff
@@ -38,6 +38,7 @@ services:
       - .:/code
   diff-sync:
     image: destructhub/project_euler
-    command: bash test
+    command:
```
Comment (ryukinix) [inline docker-compose.yml]: bash test is working since you use `-e` param of docker-compose as I did.

**Comment by A-Hilaly** [comment]
Damn , I missed everything . I made the modifications now

**Comment by A-Hilaly** [comment]
Are those errors normal ? I found out that it is possible to add envs vars in `docker-compose.yml`
 ```YAML
environment:
      - TRAVIS_PULL_REQUEST=${TRAVIS_PULL_REQUEST}
```
in `diff-sync `in `docker-compose.yaml`

**Comment by ryukinix** [comment]
Yes, I'm trying to send the env by command line but this is not working. Our last option probably is fixing that on the docker-compose.yml file, but I don't know exactly if this will work as expected.

**Comment by ryukinix** [comment]
[works badge](https://github.com/nikku/works-on-my-machine)
On my machine works:


```
❯ docker-compose run -e CANCER=CANCER diff-sync env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=38ab642a17d2
TERM=xterm
CANCER=CANCER
LANG=en_US.UTF-8
LANGUAGE=en_US:en
LC_ALL=en_US.UTF-8
HOME=/root

```

**Comment by A-Hilaly** [comment]
how about removing the first check at ./test file and just running `[ $TRAVIS_PULL_REQUEST = "false" ] ||  docker-compose run diff-sync` ? it will exit if first condition is true ( travis/push )

**Comment by ryukinix** [comment]
Seems that I find the bug @A-Hilaly

**Comment by ryukinix** [comment]
> how about removing the first check at ./test file and just running [ $TRAVIS_PULL_REQUEST = "false" ] || docker-compose run diff-sync ? it will exit if first condition is true ( travis/push )

This is similar with [something](https://github.com/DestructHub/ProjectEuler/pull/143/commits/7b79deb1cf86a4ef953e83d3d943f83932608755) that I wrote earlier... but you overwrite that changes by git pushing it.

**Comment by ryukinix** [comment]
FINALLY IS FIXED.

**Comment by ryukinix** [comment]
@A-Hilaly https://github.com/DestructHub/ProjectEuler/pull/143/commits/6feb53d143bc1528537fb55bdcabf21abb542541

This fixes. I can blame you, but relax.

**Comment by A-Hilaly** [comment]
> This fixes. I can blame you, but relax.

:owned:

**Comment by A-Hilaly** [comment]
Finally ... this was really painful , I thought that was a .yml syntax. Well done. Sorry for all the mess :owned:**2

**Comment by ryukinix** [comment]
@A-Hilaly if you are ok with this code for now, you can accept my changes by reviewing it (Approve) and all the job is done.

**Comment by ryukinix** [comment]
@A-Hilaly thanks for all the patience with our mess (your too for now). xD

**Comment by A-Hilaly** [comment]
Hahaha I missed too hard in **live** dev, my brain lagged too hard because of it . Anyways good work 👍 (I promise that the PR implementing travis commenting build results wouldn't be painful as this one xd)

**Pull request merged**